grakn/rpc/cluster/session.py

```diff
@@ -16,137 +16,68 @@
-        return selected_session.transaction(transaction_type, options)
-
-    def _fetch_latest_replica_info(self, max_retries) -> ReplicaInfo.Replica:
-        retries = 0
-        while retries < max_retries:
-            self._database = self._discover_database()
-            if self._database.primary_replica():
-                return self._database.primary_replica()
-            else:
-                time.sleep(self.WAIT_FOR_PRIMARY_REPLICA_SELECTION_SECONDS)
-                retries += 1
-        raise self._cluster_not_available_exception()
-
-    def _discover_database(self, server_address: ServerAddress = None) -> ReplicaInfo:
-        if server_address:
-            db_discover_req = database_proto.Database.Discover.Req()
-            db_discover_req.database = self._db_name
-            res = self._cluster_client.grakn_cluster_grpc_stub(server_address).database_discover(db_discover_req)
-            db = ReplicaInfo.of_proto(res)
-            print("Requested database discovery from peer %s, and got response: %s" % (str(server_address), str([str(replica) for replica in db.replicas()])))
-            return db
-        else:
-            for server_addr in self._cluster_client.cluster_members():
-                try:
-                    return self._discover_database(server_addr)
-                except RpcError as e:
-                    print("Unable to perform database discovery to %s. Attempting next address. %s" % (str(server_addr), str(e)))
-            raise self._cluster_not_available_exception()
-
-    def _cluster_not_available_exception(self) -> GraknClientException:
-        addresses = str([str(addr) for addr in self._cluster_client.cluster_members()])
-        return GraknClientException("Unable to connect to Grakn Cluster. Attempted connecting to the cluster members, but none are available: '%s'" % addresses)
+
+class TransactionFailsafeTask(FailsafeTask):
+
+    def __init__(self, cluster_session: SessionClusterRPC, transaction_type: TransactionType, options: GraknClusterOptions):
+        super().__init__(cluster_session.cluster_client, cluster_session.database())
+        self.cluster_session = cluster_session
+        self.transaction_type = transaction_type
+        self.options = options
+
+    def run(self, replica: ReplicaInfo.Replica):
+        return self.cluster_session.core_session.transaction(self.transaction_type, self.options)
+
+    def rerun(self, replica: ReplicaInfo.Replica):
+        if self.cluster_session.core_session:
+            self.cluster_session.core_session.close()
+        self.cluster_session.core_client = self.cluster_session.cluster_client.core_client(replica.address())
+        self.cluster_session.core_session = self.cluster_session.core_client.session(self.database, self.cluster_session.session_type(), self.options)
+        return self.cluster_session.core_session.transaction(self.transaction_type, self.options)
```

Comment:
In `FailsafeTask`, `run` is called on the first run, and `rerun` is called on every subsequent run, if the first run fails.

In this case, if the first run fails, we need to tell the Cluster session to:

1) close its Core session
2) switch to a different Core client (linked to a different server node)
3) try to open a new Core session in the new client
4) try to open the transaction using this Core session, instead of the original one.

This allows us to fail-over the opening of Transactions, while enforcing that each Cluster session has, at any one time, precisely one Core session associated with it.